src/transpilation/resolve.ts

```diff
@@ -0,0 +1,206 @@
+        diagnostics.push(...resolutionResult.diagnostics);
+    }
+
+    return { resolvedFiles: outFiles, diagnostics };
+}
+
+function resolveFileDependencies(file: ProcessedFile, program: ts.Program, emitHost: EmitHost): ResolutionResult {
+    const dependencies: ProcessedFile[] = [];
+    const diagnostics: ts.Diagnostic[] = [];
+
+    const projectRootDir = getSourceDir(program);
+
+    for (const required of findRequiredPaths(file.code)) {
+        // Do no resolve lualib
+        if (required === "lualib_bundle") {
+            continue;
+        }
+
+        // Do not resolve noResolution paths
+        if (required.startsWith("@NoResolution:")) {
+            const path = required.replace("@NoResolution:", "");
+            replaceRequireInCode(file, required, path);
+            replaceRequireInSourceMap(file, required, path);
+            continue;
+        }
+
+        // Try to resolve the import starting from the directory `file` is in
+        const fileDir = path.dirname(file.fileName);
+        const resolvedDependency = resolveDependency(fileDir, projectRootDir, required, emitHost);
+        if (resolvedDependency) {
+            // Figure out resolved require path and dependency output path
+            const resolvedRequire = getEmitPathRelativeToOutDir(resolvedDependency, program);
+
+            replaceRequireInCode(file, required, resolvedRequire);
+            replaceRequireInSourceMap(file, required, resolvedRequire);
+
+            // If dependency is not part of project, add dependency to output and resolve its dependencies recursively
+            if (!isProjectFile(resolvedDependency)) {
+                // If dependency resolved successfully, read its content
+                const dependencyContent = emitHost.readFile(resolvedDependency);
+                if (dependencyContent === undefined) {
+                    diagnostics.push(couldNotReadDependency(resolvedDependency));
+                    continue;
+                }
+
+                const dependency = {
+                    fileName: resolvedDependency,
+                    code: dependencyContent,
+                };
+                const nestedDependencies = resolveFileDependencies(dependency, program, emitHost);
+                dependencies.push(dependency, ...nestedDependencies.resolvedFiles);
+                diagnostics.push(...nestedDependencies.diagnostics);
+            }
+        } else {
+            // Could not resolve dependency, add a diagnostic and make some fallback path
+            diagnostics.push(couldNotResolveRequire(required, path.relative(projectRootDir, file.fileName)));
+
+            const fallbackRequire = fallbackResolve(required, projectRootDir, fileDir);
+            replaceRequireInCode(file, required, fallbackRequire);
+            replaceRequireInSourceMap(file, required, fallbackRequire);
+        }
+    }
+    return { resolvedFiles: dependencies, diagnostics };
+}
```

Comment:
This function is a bit long and contains to many nested conditionals. Would be nice to split it into multiple functions, that might also make some of the comments obsolete.

Reply:
I got it as short as I could now, it's still quite long but any further separation just hurts readability imo.